File: acmOtherMaterials/Ulm Local Contest/2003/solution/genetic.pas
(* Problem   Genetic Code
** Algorithm Backtracking
** Runtime   O(3^n)
** Author    Walter Guttmann
** Date      12.03.2003
*)

program genetic;

var f: Text;
    s: array [0..8191] of Char;
    i, n: Longint;

function strncmp(i1, i2, l: Longint): Longint;
begin
  while l > 0 do begin
    if s[i1] < s[i2] then begin
      strncmp := -1;
      exit
    end else if s[i1] > s[i2] then begin
      strncmp := 1;
      exit
    end;
    i1 := i1 + 1;
    i2 := i2 + 1;
    l := l - 1
  end;
  strncmp := 0
end;

function isThue(e: Longint): Boolean;
var len: Longint;
begin
  isThue := False;
  for len := 1 to e div 2 do
    if strncmp(e-len-len, e-len, len) = 0 then
      exit;
  isThue := True;
end;

function backtrack(e: Longint): Boolean;
var ch: Char;
begin
  backtrack := True;
  if e = 5000 then
    exit;
  for ch := 'N' to 'P' do begin
    s[e] := ch;
    if isThue(e) then
      if backtrack(e+1) then
        exit
  end;
  backtrack := False
end;

begin
  backtrack(0);
  Assign(f, 'genetic.in');
  Reset(f);
  Read(f, n);
  while n <> 0 do begin
    for i := 0 to n-1 do
      Write(s[i]);
    WriteLn();
    Read(f, n)
  end;
  Close(f)
end.

File: acmOtherMaterials/Ulm Local Contest/2000/judge's solution/genetic.cc
// Problem   Phylogenetic Trees Inherited
// Algorithm Dynamic Programming
// Runtime   O(n * l)
// Author    Walter Guttmann
// Date      12.01.2000

#include <fstream>

ifstream in ("genetic.in");

static char seq[1024][1024];

main ()
{
  int heap[2048], n, l;
  while (in >> n >> l)
  {
    if (n == 0 || l == 0) break;
    for (int k=0 ; k<n ; k++)
      in >> seq[k];
    int cost = 0;
    for (int i=0 ; i<l ; i++)
    {
      for (int k=0 ; k<n ; k++)
        heap[n+k] = 1 << (seq[k][i] - 'A');
      for (int k=n-1 ; k>0 ; k--)
        if (! (heap[k] = heap[k+k] & heap[k+k+1]))
          ++cost, heap[k] = heap[k+k] | heap[k+k+1];
      char ch = 'A';
      while (heap[1]>>=1) ++ch;
      cout << ch;
    }
    cout << ' ' << cost << endl;
  }
  return 0;
}

